src/open_inwoner/react/lib/web-component/loader.tsx

```diff
@@ -0,0 +1,186 @@
+   * @param error The error that occured.
+   */
+  private static async runErrorHooks(
+    context: WebComponentContext,
+    error: Error
+  ) {
+    if (!error) return;
+
+    for (const plugin of WebComponentLoader.pluginRegistry) {
+      try {
+        plugin.onError?.(context, error);
+      } catch (pluginError) {
+        console.error(
+          `[Plugin:${plugin.name}] onError hook failed for ${context.componentName}:`,
+          pluginError
+        );
+      }
+    }
+  }
+
+  private static registerWebComponent<P = {}>(
+    Component: FC<P>,
+    tagName: string,
+    propNames?: (keyof P)[],
+    options: WebComponentRegisterOptions = { shadow: false, i18n: false }
+  ): HTMLElement {
+    // Remove i18n from options before passing to preact-custom-element
+    const { i18n, ...registerOptions } = options;
+
+    // If i18n option is enabled, wrap the component with IntlProvider
+    const ComponentToRegister = i18n ? withIntlWc(Component) : Component;
+
+    return register(ComponentToRegister, tagName, propNames, registerOptions);
+  }
+
+  public static async importWC(name: WebComponentTagName): Promise<void> {
+    // Skip if already defined
+    if (customElements.get(name))
+      return console.debug(
+        `Custom element ${name} is already defined, aborting loading for this element`
+      );
+
+    // Get the import function.
+    const { importer } = WebComponentLoader.registry[name];
+    if (!importer) throw new Error(`"${name}" has no web component import`);
+
+    // Get the context of the web components
+    const contexts = WebComponentLoader.createContextsForComponent(name);
+
+    try {
+      // Run plugins `beforeLoad` hooks
+      await Promise.all(contexts.map(WebComponentLoader.runBeforeLoadHooks));
+
+      // Import and load the web component
+      const { default: Component } = await importer();
+      if (!Component) throw new Error(`"${name}" has no default export`);
+      const config = WebComponentLoader.registry[name];
+      type ExtractGeneric<Type> = Type extends FC<infer P> ? P : never;
+
+      Component as FC<ExtractGeneric<typeof Component>>;
+
+      WebComponentLoader.registerWebComponent<ExtractGeneric<typeof Component>>(
+        Component as FC<ExtractGeneric<typeof Component>>,
+        config.tagName,
+        config.propNames as (keyof ExtractGeneric<typeof Component>)[],
+        config.options
+      );
+
+      // Run plugins `afterLoad` hooks
+      await Promise.all(contexts.map(WebComponentLoader.runAfterLoadHooks));
+    } catch (err) {
+      // Run plugins `onError` hooks
+      contexts.forEach((context) =>
+        WebComponentLoader.runErrorHooks(context, err as Error)
+      );
+    }
+  }
+
+  /**
+   * Register the webcomponents on the current page
+   * @returns
+   */
+  static async registerWebComponents() {
+    try {
+      // Look for all web components on the current page
+      const founded = WebComponentLoader.findWebComponentsOnPage();
```

Comment:
Maybe also instantiate `WebComponentLoader`? Now everything in the class is static, but there's no guarantee it'll stay that way, and the overhead is minimal.


```suggestion
      const founded = WebComponentLoader().findWebComponentsOnPage();
```

Reply:
For now i'll leave it like this (static), we can always refactor it later